adapter/utils/src/main/java/org/mobilitydata/gtfsvalidator/geoutils/GeospatialUtilsImpl.java

```diff
@@ -63,8 +79,72 @@ public int distanceBetweenMeter(double fromLat, double fromLng, double toLat, do
                 * KILOMETER_TO_METER_CONVERSION_FACTOR);
     }
 
+    /**
+     * Returns a list of E047 errors for the given input, one for each stop that is too far from the trip shapePoints
+     *
+     * @param trip        Trip for this GTFS trip
+     * @param stopTimes   a map of StopTimes for a trip, sorted by stop_sequence
+     * @param shapePoints a map of ShapePoints for a trip, sorted by shape_pt_sequence
+     * @param stops       a map of all stops (keyed on stop_id), needed to obtain the latitude and longitude for each stop
+     * @param testedCache a cache for previously tested shape_id and stop_id pairs (keyed on shape_id+stop_id). If the
+     *                    combination of shape_id and stop_id appears in this set, we shouldn't test it again. Shapes
+     *                    and stops tested in this method execution will be added to this testedCache.
+     * @return a list of E047 errors, one for each stop that is too far from the trip shapePoints
+     */
+    public List<StopTooFarFromTripShape> checkStopsWithinTripShape(Trip trip,
+                                                                   SortedMap<Integer, StopTime> stopTimes,
+                                                                   SortedMap<Integer, ShapePoint> shapePoints,
+                                                                   Map<String, LocationBase> stops,
+                                                                   Set<String> testedCache) {
+        List<StopTooFarFromTripShape> errors = new ArrayList<>();
+        if (trip == null || stopTimes == null || stopTimes.isEmpty() || shapePoints == null || shapePoints.isEmpty()) {
+            // Nothing to do - return empty list
+            return errors;
+        }
+
+        // Create a polyline from the GTFS shapes data
+        ShapeFactory sf = JtsSpatialContext.GEO.getShapeFactory();
+        ShapeFactory.LineStringBuilder lineBuilder = sf.lineString();
+        shapePoints.forEach((integer, shapePoint) -> lineBuilder.pointXY(shapePoint.getShapePtLon(), shapePoint.getShapePtLat()));
+        Shape shapeLine = lineBuilder.build();
+
+        // Create the buffered version of the trip as a polygon
+        Shape shapeBuffer = shapeLine.getBuffered(TRIP_BUFFER_DEGREES, shapeLine.getContext());
+
+        // Check if each stop is within the buffer polygon
+        stopTimes.forEach((integer, stopTime) -> {
+            LocationBase stop = stops.get(stopTime.getStopId());
+            if (stop == null || stop.getStopLat() == null || stop.getStopLon() == null) {
+                // Lat/lon are optional for location_type 4 - skip to the next stop if they aren't provided
+                return;
+            }
+            if (!(stop instanceof StopOrPlatform) && !(stop instanceof BoardingArea)) {
+                // This rule only applies to stops of location_type 0 and 4 - skip to next stop
```

Comment:
I believe this rule only applies to location_types 0 and 4, but I've asked the spec team to clarify.

Reply:
This is correct per spec team.